feat: 예약 신청/취소 기능 구현

jeongyeon0208 opened the pull request:
## 🌱 관련 이슈

- [LCR-103]

---
## 📌 작업 내용 및 특이사항

- MemberReservation 테이블에 예약의 상태를 저장할 status, 입장 여부를 확인할 isEntered 속성을 추가했습니다. (+flyway 스크립트)
- MemberServiceClient(FeignClient)를 통해 findMemberInfo를 호출하여 예약을 하는 회원 정보를 받아옵니다.

- MemberReservation 생성 메서드를 구현했습니다. 
  - 중복 예약 검증을 통과한 후 Redis에서 reservationId를 키값으로 하는 데이터를 찾아서 1 감소시킵니다.
    -  Redis에서 오류가 나서 감소 실패했을 시 RedisConnectionFailureException를 발생시키고 메서드가 종료됩니다.
    - Redis 감소가 성공적으로 이루어졌으면, 차감 이후 값을 통해 예약 가능 여부를 확인합니다.
    - Redis 차감 이후 값이 없거나 '0' 미만인 경우, 다시 Redis에서 값을 증가시킨 후 예약 실패 예외를 던집니다.
  - Redis 차감 이후 예약이 가능한 경우 memberId, reservationId 정보만 포함된 MemberReservation을 생성후 저장합니다. 
    - 이때 defalut status는 "PENDING"입니다.
  - 위의 MemberReservation 저장 과정에서 예외가 발생한다면, Redis에서 다시 값을 증가시키고 예외를 던집니다.
  - MemberReservation 저장까지 성공하면 MemberReservation의 id를 담은 MemberReservationUpdateEvent를 발행합니다.
- 이벤트 핸들러는 MemberReservationUpdateEvent를 감지하여 비동기로, MemberReservation 저장 커밋 이후 실행됩니다.
  - 이벤트를 처리하면서 updateMemberReservation 메서드를 실행하고 이때 실패하면 재시도합니다. 이때 재시도 시간은 1초 후, 2초 후, 4초 후...  최대 9번 실행합니다. 하지만 만약 메서드 실패의 원인이 MemberReservation, Reservation의 부재라면 더이상 재시도 하지 않고 return 합니다.

  - 최대 9번 재시도 끝에도 업데이트 성공하지 못하면 MemberReservation를 삭제합니다.
- MemberReservation 업데이트 메서드를 구현했습니다. 
  - 이벤트 핸들러를 통해 호출됩니다.
  - 이벤트 DTO에 있는 memberReservationId로 조회합니다.
  - 존재하는 MemberReservation이면 ManagerServiceClient(FeighnClient)를 통해 팝업과 예약 정보를 받아 MemberReservation을 업데이트 하고 상태를 "RESERVED"로 변경합니다.

- Redis 증가는 실패하면 안되므로 Resilience4j를 이용해서 increment 연산에 대해 안정성을 확보했습니다.
  - safeIncrement() 메서드는 Redis의 특정 key 값을 1 증가시키는 기능을 수행하며, Resilience4j의 Circuit Breaker + Retry를 적용했습니다.
  - Redis increment 중에 RuntimeException이 발생하면 최대 5번까지 200ms 간격으로 재시도합니다.
  - 최근 10번의 호출 중 50% 이상이 실패하면 circuit이 열림(OPEN) 상태가 됩니다.
  - 최소 5번 이상 호출되어야 failure rate 체크 시작합니다.
  - OPEN 상태일 경우, 10초간 차단 후 반쯤 열림(HALF-OPEN) 상태로 자동 전환합니다.
  - HALF-OPEN 상태에선 2번만 시도해보고 성공 여부에 따라 다시 CLOSED or OPEN 결정합니다.
  - Redis가 자꾸 실패하면 더 이상 호출하지 않고 차단하며, 일정 시간 후 회복을 자동으로 시도합니다.
  - Circuit Breaker가 OPEN 상태이거나 모든 재시도 실패 시, fallback 메서드가 호출되어 실패한 키와 에러 메시지를 로그로 남깁니다.

- 회원 예약 취소 API 구현했습니다.

---
## 📚 참고사항

- API 명세서에 각 실패 상황에 대한 응답 코드 작성했습니다. [API 명세서](https://www.notion.so/API-1e1483a29ee480949f5be94fd080e72b?p=1f8483a29ee4809e86feee3a882d4cc3&pm=s)


[LCR-103]: https://lgcns-retail.atlassian.net/browse/LCR-103?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jeongyeon0208 committed:
feat: MemberReservation 테이블에 status, isEntered 속성 추가

jeongyeon0208 committed:
feat: 회원 검증용 MemberServiceClient 구현

# Conflicts:
#	popi-reservation-service/src/main/java/com/lgcns/client/managerClient/dto/response/DailyReservation.java
#	popi-reservation-service/src/main/java/com/lgcns/client/managerClient/dto/response/MonthlyReservationResponse.java
#	popi-reservation-service/src/main/java/com/lgcns/client/managerClient/dto/response/TimeSlot.java

jeongyeon0208 committed:
feat: MemberReservation 생성 메서드 구현

# Conflicts:
#	popi-reservation-service/src/main/java/com/lgcns/service/MemberReservationService.java
#	popi-reservation-service/src/main/java/com/lgcns/service/MemberReservationServiceImpl.java

jeongyeon0208 committed:
feat: ManagerServiceClient에 예약 정보 조회 추가

jeongyeon0208 committed:
feat: memberReservation update event handler 구현

jeongyeon0208 committed:
chore: qr, circuit breaker 의존성 추가

jeongyeon0208 committed:
refactor: RedisConfig에 redisTemplate 추가

jeongyeon0208 committed:
feat: 회원예약 검증 메서드 추가

jeongyeon0208 committed:
feat: 회원 예약 생성 api 구현

jeongyeon0208 committed:
refactor: ManagerServiceClient 내 메서드 명 수정

jeongyeon0208 committed:
chore: DatabaseCleaner 추가

jeongyeon0208 committed:
test: createMemberReservation 서비스 테스트 작성

jeongyeon0208 committed:
test: createMemberReservation 실패 Mock 테스트 작성

jeongyeon0208 committed:
rename: 이벤트 메서드 변경

jeongyeon0208 committed:
refactor: reservation api url 내 popups 제거

jeongyeon0208 committed:
refactor: ddl-auto validate 설정

jeongyeon0208 committed:
refactor: REDIS_HOST 기본값 localhost 설정

jeongyeon0208 committed:
test: calceMemberReservation 테스트 코드 작성

jeongyeon0208 committed:
test: updateMemberReservation 테스트 코드 작성

0ssang reviewed (approved):
비동기 이벤트 처리 방법과 Retry, CircuitBreaker 사용법 등 많이 배워갑니다!!
LGTM 👍👍

ohhyungsuh commented on popi-reservation-service/src/main/java/com/lgcns/domain/MemberReservation.java:
```diff
@@ -20,17 +20,19 @@ public class MemberReservation extends BaseTimeEntity {
+    @Enumerated(EnumType.STRING)
+    private MemberReservationStatus status = MemberReservationStatus.PENDING;
+
+    private Boolean isEntered = false;
```
이거 Builder에 넣으면 가독성이 더 좋을 것 같아요!

jeongyeon0208 replied:
수정했습니다!

jeongyeon0208 committed:
refactor: MemberReservation default 속성 위치 수정

jeongyeon0208 committed:
refactor: 미사용 ErrorCode 제거

ohhyungsuh commented on popi-reservation-service/src/main/java/com/lgcns/service/MemberReservationServiceImpl.java:
```diff
@@ -63,6 +174,52 @@ private void validateYearMonthFormat(String date) {
+            data.put("memberId", memberId);
+            data.put("popupId", popupId);
+            data.put("reservationDate", reservationDate.toString());
+            data.put("reservationTime", reservationTime.toString());
```
id대신 memberReservationId, memberId대신 성별, 나이대가 필요할 것 같아요!

jeongyeon0208 replied:
수정했습니다~

jeongyeon0208 committed:
refactor: /reservation/popups 엔드포인트 복구

jeongyeon0208 committed:
refactor: MemberInternalInfoResponse 수정

# Conflicts:
#	popi-common/src/main/java/com/lgcns/dto/response/MemberInternalInfoResponse.java
#	popi-member-service/src/main/java/com/lgcns/service/MemberServiceImpl.java

jeongyeon0208 committed:
refactor: MemberReservation CANCELED 상태 제거

jeongyeon0208 committed:
refactor: MemberInternalInfoResponse stub 수정

jeongyeon0208 committed:
refactor: 예약 시 멤버 재검증 제거

jeongyeon0208 committed:
feat: 예약 취소 API 구현

jeongyeon0208 committed:
refactor: 예약QR 저장 정보 수정

jeongyeon0208 committed:
refactor: DTO import

jeongyeon0208 committed:
style: spotless

jeongyeon0208 committed:
refactor: PopupIdsRequest 적용

ht3064 commented on popi-reservation-service/src/main/java/com/lgcns/PopiReservationServiceApplication.java:
```diff
@@ -2,8 +2,10 @@
+import org.springframework.scheduling.annotation.EnableAsync;
 
 @SpringBootApplication
+@EnableAsync
```
해당 @EnableAsync 설정은 AsyncConfig 같은 별도 설정 클래스로 분리하면 좋을 것 같아요!

jeongyeon0208 replied:
넵 추후 이벤트 핸들러 테스트 코드 진행하면서 리팩토링 진행하겠습니다!

jeongyeon0208 committed:
refactor: /popups/... endpoint 복구

ht3064 commented on popi-reservation-service/src/main/java/com/lgcns/service/MemberReservationServiceImpl.java:
```diff
@@ -100,6 +117,89 @@ public List<Long> findHotPopupIds() {
+        try {
+            memberReservation =
+                    MemberReservation.createMemberReservation(
+                            reservationId, Long.parseLong(memberId), null, null, null, null);
```
이 부분은 정적 팩토리 메서드에서 null을 받지 않도록 하고 사용하지 않는 파라미터는 제거하면 좋을 것 같아요!

jeongyeon0208 replied:
수정했습니다!

woogieon8on reviewed (approved):
너무 고생하셨습니다! 예약 구현할 때 고려해야 할 사항이 생각보다도 훨씬 더 많네요!
LGTMMMMMMMMMMMMM!!!

ht3064 commented on popi-reservation-service/src/main/java/com/lgcns/service/MemberReservationServiceImpl.java:
```diff
@@ -100,6 +117,89 @@ public List<Long> findHotPopupIds() {
+                                (() ->
+                                        new CustomException(
+                                                MemberReservationErrorCode
+                                                        .MEMBER_RESERVATION_NOT_FOUND)));
```
이 부분 중복되니까 메서드로 따로 빼면 좋을 듯 해요!

jeongyeon0208 replied:
이부분 메서드로 추출완료했습니다.

jeongyeon0208 committed:
refactor: MemberReservation 내부 생성 메서드 수정

jeongyeon0208 committed:
refactor: findMemberReservation 메서드 추출

ht3064 reviewed (approved):
서킷브레이커 사용까지 하시고 고생많으셨습니다! LGTM 👍
